### ISSUE TITLE
CT: coin term is already present in expression

### ISSUE THREAD

**Issue by jbphet** [phetsims/expression-exchange#151]
I made a number of changes to the drag handlers recently to try to fix multi-touch CT errors, but it looks like a new one may have cropped up.  Here is the stack trace:

```
expression-exchange : pan-and-zoom-fuzz : unbuilt
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1615064572807/expression-exchange/expression-exchange_en.html?continuousTest=%7B%22test%22%3A%5B%22expression-exchange%22%2C%22pan-and-zoom-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1615064572807%22%2C%22timestamp%22%3A1615128198952%7D&brand=phet&ea&fuzz&fuzzPointers=2&memoryLimit=1000&supportsPanAndZoom=true
Query: brand=phet&ea&fuzz&fuzzPointers=2&memoryLimit=1000&supportsPanAndZoom=true
Uncaught Error: Assertion failed: coin term is already present in expression
Error: Assertion failed: coin term is already present in expression
at window.assertions.assertFunction (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1615064572807/assert/js/assert.js:25:13)
at Expression.addCoinTerm (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1615064572807/expression-exchange/js/common/model/Expression.js:405:15)
at coinTermUserControlledListener (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1615064572807/expression-exchange/js/common/model/ExpressionManipulationModel.js:779:37)
at TinyEmitter.emit (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1615064572807/axon/js/TinyEmitter.js:86:18)
at Property._notifyListeners (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1615064572807/axon/js/Property.js:294:25)
at Property.set (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1615064572807/axon/js/Property.js:191:14)
at DragListener.end [as _end] (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1615064572807/expression-exchange/js/common/view/AbstractCoinTermNode.js:374:46)
at https://bayes.colorado.edu/continuous-testing/ct-snapshots/1615064572807/scenery/js/listeners/DragListener.js:337:25
at DragListener.onRelease (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1615064572807/scenery/js/listeners/PressListener.js:646:17)
at Action.execute (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1615064572807/axon/js/Action.js:225:18)
id: Bayes Chrome
Snapshot from 3/6/2021, 2:02:52 PM
```

**Comment by jbphet** [comment]
I was able to duplicate this by running fuzz testing, but it took a while.  In an effort to capture exactly what test was being run, here is the URL that I used:

```
http://localhost:8080/expression-exchange/expression-exchange_en.html?&brand=phet&ea&fuzz&fuzzPointers=2&memoryLimit=1000&supportsPanAndZoom=true&debugger
```

The stack trace looks pretty much exactly like the one that was captured during CT, so I'm pretty sure that it's the same situation.  The trace is below in case I need it for reference.

<details>
<summary>Stack trace</summary>

```
window.assertions.assertFunction (assert.js:23)
addCoinTerm (Expression.js:405)
coinTermUserControlledListener (ExpressionManipulationModel.js:779)
emit (TinyEmitter.js:86)
_notifyListeners (Property.js:294)
set (Property.js:191)
end (AbstractCoinTermNode.js:374)
(anonymous) (DragListener.js:337)
onRelease (PressListener.js:646)
execute (Action.js:225)
release (PressListener.js:427)
release (DragListener.js:333)
pointerUp (PressListener.js:757)
dispatchToListeners (Input.js:1842)
dispatchEvent (Input.js:1791)
upEvent (Input.js:1571)
Input.touchEndAction.Action.phetioPlayback (Input.js:415)
execute (Action.js:225)
touchEnd (Input.js:1218)
touchEnd (InputFuzzer.js:233)
InputFuzzer.touchEndAction (InputFuzzer.js:55)
fuzzEvents (InputFuzzer.js:105)
fuzzInputEvents (Sim.js:1084)
runAnimationLoop (Sim.js:1035)
requestAnimationFrame (async)
.
.
.
```

</details>

The basic story that is being told by this stack trace is that a coin term was being dragged, and it was released over an expression, so it should be added to that expression.  But wait - the expression thinks that this coin term is _already_ in this expression.  That shouldn't happen, since, by design, you shouldn't be able to drag a coin term by itself when it's part of an expression unless the expression is in edit mode.  Boom (throw an assertion).

**Comment by jbphet** [comment]
I think I have a fix for this, but it's hard to be sure because it only happened rarely in the first place, and it's difficult to say for certain that it's gone.  I tested the change locally running the fuzz test for 40 minutes with no sign of the problem, so I plan to keep an eye on CT for a couple of days and, if it doesn't show up again, I'll declare it fixed and close this issue.

The solution, in a nutshell, was to interrupt dragging interactions with a coin term node when the coin term gets broken from a collective set to a group of individuals.  In my testing and debugging, this was the only situation in which the coin term was in a user-controlled state when it was supposed to be non-pickable.  I was never able to duplicate the problem manually, but I added some timer output to the console and found that each time the assertion was hit the non-pickable-while-user-controlled state happened just before it.  Almost a smoking gun, but not quite.

**Comment by jbphet** [comment]
CT is clear after about 19 hours.  I also did a bunch more local fuzz testing with the query parameters listed in the original problem report, and did not hit any assertions.  I'm going to go ahead and close this, but I'll check CT every work day (as I generally do) and will re-open if I see the same problem again.

**Comment by KatieWoe** [comment]
This is showing up in CT again:
```
expression-exchange : multitouch-fuzz : unbuilt
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1618459372043/expression-exchange/expression-exchange_en.html?continuousTest=%7B%22test%22%3A%5B%22expression-exchange%22%2C%22multitouch-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1618459372043%22%2C%22timestamp%22%3A1618466745872%7D&brand=phet&ea&fuzz&fuzzPointers=2&memoryLimit=1000&supportsPanAndZoom=false
Query: brand=phet&ea&fuzz&fuzzPointers=2&memoryLimit=1000&supportsPanAndZoom=false
Uncaught Error: Assertion failed: coin term is already present in expression
Error: Assertion failed: coin term is already present in expression
at window.assertions.assertFunction (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1618459372043/assert/js/assert.js:25:13)
at Expression.addCoinTerm (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1618459372043/expression-exchange/js/common/model/Expression.js:405:15)
at coinTermUserControlledListener (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1618459372043/expression-exchange/js/common/model/ExpressionManipulationModel.js:779:37)
at TinyProperty.emit (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1618459372043/axon/js/TinyEmitter.js:86:18)
at Property._notifyListeners (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1618459372043/axon/js/Property.js:269:23)
at Property.set (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1618459372043/axon/js/Property.js:184:14)
at DragListener.end [as _end] (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1618459372043/expression-exchange/js/common/view/AbstractCoinTermNode.js:370:46)
at https://bayes.colorado.edu/continuous-testing/ct-snapshots/1618459372043/scenery/js/listeners/DragListener.js:337:25
at DragListener.onRelease (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1618459372043/scenery/js/listeners/PressListener.js:646:17)
at Action.execute (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1618459372043/axon/js/Action.js:227:18)
id: Bayes Chrome
Snapshot from 4/14/2021, 10:02:52 PM
```

**Comment by jbphet** [comment]
Okay, with much effort, I have finally figured out how to duplicate this.  It requires a touch interface.  Here is the sequence:

1. Go to any of the first three screens
2. Create an expression with two or more coins/terms in it
3. Set the expression to edit mode by clicking it, then pressing the edit button (the white one with the arrows on it)
4. Press and hold somewhere on the barrier rectangle (the gray rectangle that prevents interaction with anything outside of the expression) outside of the expression
5. With another finger, and while continuing to hold down the finger on the barrier rectangle, drag a coin/term out of the expression and hold it there
6. Release the finger that was on the barrier rectangle
7. Drag the coin/term back to the expression and release it

**Comment by jbphet** [comment]
I think I have a fix for this, see the commit above.  In summary, I've made it so that there is no "arming" step for getting rid of the barrier rectangle when editing expressions.  This is necessary because if an arming step is used, multi-touch situations can cause the barrier rectangle to go away when it shouldn't.

I worked on #152 at the same time and ran fuzz testing locally for a little over 1/2 hr and saw no problems.  I'll call this fixed, but will watch CT and will reopen if I see it again.  @KatieWoe - you're welcome to do the same.
